sdk/export/metric/metric.go

```diff
@@ -21,70 +21,271 @@ import (
+	// metric instrument.  This must return a consistent type to
+	// avoid confusion in later stages of the metrics export
+	// process, i.e., when Merging multiple aggregators for a
+	// specific instrument.
+	//
+	// Note: This is context-free because the aggregator should
+	// not relate to the incoming context.  This call should not
+	// block.
+	AggregatorFor(*Descriptor) Aggregator
 }
 
-// Aggregator implements a specific aggregation behavior, e.g.,
-// a counter, a gauge, a histogram.
+// Aggregator implements a specific aggregation behavior, e.g., a
+// behavior to track a sequence of updates to a counter, a gauge, or a
+// measure instrument.  For the most part, counter and gauge semantics
+// are fixed and the provided implementations should be used.  Measure
+// metrics offer a wide range of potential tradeoffs and several
+// implementations are provided.
+//
+// Note that any Aggregator may be attached to any instrument--this is
+// the result of the OpenTelemetry API/SDK separation.  It is possible
+// to attach a counter aggregator to a measure instrument (to compute
+// a simple sum) or a gauge instrument to a measure instrument (to
+// compute the last value).
 type Aggregator interface {
 	// Update receives a new measured value and incorporates it
-	// into the aggregation.
-	Update(context.Context, core.Number, Record)
+	// into the aggregation.  Update() calls may arrive
+	// concurrently as the SDK does not provide synchronization.
+	//
+	// Descriptor.NumberKind() should be consulted to determine
+	// whether the provided number is an int64 or float64.
+	//
+	// The Context argument comes from user-level code and could be
+	// inspected for distributed or span context.
+	Update(context.Context, core.Number, *Descriptor) error
 
-	// Collect is called during the SDK Collect() to
-	// finish one period of aggregation.  Collect() is
-	// called in a single-threaded context.  Update()
-	// calls may arrive concurrently.
-	Collect(context.Context, Record, Batcher)
+	// Checkpoint is called in collection context to finish one
+	// period of aggregation.  Checkpoint() is called in a
+	// single-threaded context, no locking is required.
+	//
+	// The Context argument originates from the controller that
+	// orchestrates collection.
+	Checkpoint(context.Context, *Descriptor)
 
-	// Merge combines state from two aggregators into one.
-	Merge(Aggregator, *Descriptor)
+	// Merge combines state from the argument aggregator into this
+	// one.  Merge() is called in a single-threaded context, no
+	// locking is required.
+	Merge(Aggregator, *Descriptor) error
 }
 
-// Record is the unit of export, pairing a metric
-// instrument and set of labels.
-type Record interface {
-	// Descriptor() describes the metric instrument.
-	Descriptor() *Descriptor
+// Exporter handles presentation of the checkpoint of aggregate
+// metrics.  This is the final stage of a metrics export pipeline,
+// where metric data are formatted for a specific system.
```

Comment:
Hm, how will this work for certain exporters that have different aggregation mechanisms? I'd be concerned that without letting them have raw exports for every single sample/call to inc, etc that it's not possible to actually act as they do with their native client libraries.

i.e. some exporters keep a summary of the last N minutes for quantiles, etc.

Reply:
I think certain implementors would simply need to implement "Batcher" themselves to achieve this, which may defeat the purpose of keeping batcher/exporter as different concepts. (just a concern I wanted to highlight)